go/vt/vttablet/tabletmanager/shard_sync_test.go

```diff
@@ -67,36 +68,66 @@ func TestShardSync(t *testing.T) {
 	// wait for syncing to work correctly
 	// this should also have updated the shard record since it is a more recent operation
 	// We check here that the shard record and the tablet record are in sync
-	checkShardRecordWithTimeout(ctx, t, ts, ti.Alias, ti.PrimaryTermStartTime, 1*time.Second)
+	err = checkShardRecordWithTimeout(ctx, t, ts, ti.Alias, ti.PrimaryTermStartTime, 1*time.Second)
+	require.NoError(t, err)
+
+	// Shard sync loop runs asynchronously and starts a watch on the shard.
+	// We wait for the shard watch to start, otherwise the test is flaky
+	// because the update of the record can happen before the watch is started.
+	waitForShardWatchToStart(ctx, t, tm, originalTime, ti, 10*time.Second)
 
 	// even if try to update the shard record with the old timestamp, it should be reverted again
 	updatePrimaryInfoInShardRecord(ctx, t, tm, nil, originalTime)
 
 	// this should have also updated the shard record because of the timestamp.
-	checkShardRecordWithTimeout(ctx, t, ts, ti.Alias, ti.PrimaryTermStartTime, 1*time.Second)
+	err = checkShardRecordWithTimeout(ctx, t, ts, ti.Alias, ti.PrimaryTermStartTime, 1*time.Second)
+	require.NoError(t, err)
 
 	// updating the shard record with the latest time should trigger an update in the tablet
 	newTime := time.Now()
 	updatePrimaryInfoInShardRecord(ctx, t, tm, nil, newTime)
 
 	// this should not have updated.
-	checkShardRecordWithTimeout(ctx, t, ts, nil, protoutil.TimeToProto(newTime), 1*time.Second)
+	err = checkShardRecordWithTimeout(ctx, t, ts, nil, protoutil.TimeToProto(newTime), 1*time.Second)
+	require.NoError(t, err)
 
 	// verify that the tablet record has been updated
 	checkTabletRecordWithTimeout(ctx, t, ts, tm.tabletAlias, topodata.TabletType_REPLICA, nil, 1*time.Second)
 }
 
-func checkShardRecordWithTimeout(ctx context.Context, t *testing.T, ts *topo.Server, tabletAlias *topodata.TabletAlias, expectedStartTime *vttime.Time, timeToWait time.Duration) {
+// waitForShardWatchToStart waits for shard watch to have started.
+func waitForShardWatchToStart(ctx context.Context, t *testing.T, tm *TabletManager, originalTime time.Time, ti *topo.TabletInfo, timeToWait time.Duration) {
+	// We wait for shard watch to start by
+	// updating the record and waiting to see
+	// the shard record is updated back by the tablet manager.
+	timeOut := time.After(timeToWait)
+	idx := 1
+	for {
+		select {
+		case <-timeOut:
+			t.Fatalf("timed out: waiting for shard watch to start")
```

Comment:
It's worth using `require.FailNow` here so that we have nice backtraces etc. and it's much easier to debug.